notes/2025-xx-xx-insecurities-remedies-i.md

```diff
@@ -0,0 +1,410 @@
+try and harden individual services.
+
+By default, Systemd leaves running services in the open. This is understandable,
+as trying to preemptively harden a service is likely to cause conflicts with
+individual services and their requirements.
+
+Although NixOS makes _some_ effort to harden services, installed services--be it
+through NixOS' services or your inferior distribution's package manager--will be
+running with little to no hardening - which you can confirm by viewing the
+security report. This post, as per my experience, aims to detail potential
+hardening options. Throughout this post, please keep those following core
+principles in mind:
+
+1. There is **no** "one size fits all" kind of hardening, all services will
+   require your undivided attention to make sure everything continues to work as
+   intended.
+2. No kind of hardening can catch all kinds of exploits. The key to a secure
+   system is to remain ever-vigilant.
+
+> Some hardening options will disable access to certain paths, or make them
+> read-only for the service. This is quite helpful, in theory, but not every
+> program is written intelligently and as such, not all programs will fail
+> gracefully when they are missing access to a path. Sometimes the service will
+> fail, and you will not be able to tell why. This is exactly why you must treat
+> each service with special care - and harden them one service at a time instead
+> of abstracting a generalized way of modifying multiple services at once.
+
+### Hardening Services
+
+[NixOS option search]: https://search.nixos.org/options?channel=unstable&size=50&sort=relevance&type=packages&query=systemd.services.
+
+Systemd services in NixOS are defined through the systemd module exposed in the
+Nixpkgs module system. The schema is very basic, and I suggest that you consult
+[NixOS option search] if you wish to learn more. For our purposes, just
+`serviceConfig` is enough as we will be working primarily with the `[Service]`
+field of systemd services. Documentation is very scattered, but available at
+`systemd.unit(5)`, `systemd.service(5)` and `systemd.exec(5)` manpages. They
+contain many important tidbids, and I encourage you to go through them alongside
+this post during your hardening journey.
+
+Here is a very basic example to demonstrate how individual options are applied
+to a service.
+
+```nix
+{
+  systemd.services."<serviceName>".serviceConfig = {
+    ProtectClock = true;
+    ProtectKernelTunables = true;
+    ProtectKernelModules = true;
+    ProtectKernelLogs = true;
+    SystemCallFilter = "~@clock @cpu-emulation @debug @obsolete @module @mount @raw-io @reboot @swap";
+    ProtectControlGroups = true;
+    RestrictNamespaces = true;
+    LockPersonality = true;
+    MemoryDenyWriteExecute = true;
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+  };
+}
+```
+
+This is usually enough to bring a service down to MEDIUM exposure level. It
```

Comment:
This first sentence is a little long and makes it a bit har to read. I'd break it up into separate sentences. I had to make a few double takes when reading this part.